mc/query/compile.go

```diff
@@ -535,3 +549,59 @@ func isStatementCriteria(c QueryCriteria) bool {
 		return false
 	}
 }
+
+func isIndexCriteria(c QueryCriteria) bool {
+	switch c := c.(type) {
+	case *IndexCriteria:
+		return true
+
+	case *CompoundCriteria:
+		return isIndexCriteria(c.left) || isIndexCriteria(c.right)
+
+	case *NegatedCriteria:
+		return isIndexCriteria(c.e)
+
+	default:
+		return false
+	}
+}
+
+func indexCriteriaTables(c QueryCriteria) (map[string]string, error) {
+	tabs := make(map[string]string)
+
+	err := collectIndexCriteriaTables(tabs, c)
+	if err != nil {
+		return nil, err
+	}
+
+	return tabs, err
+}
+
+func collectIndexCriteriaTables(tabs map[string]string, c QueryCriteria) error {
```

Comment:
can we call this something like `ensure...` or is `collect...` a standard prefix?